run-repeatedly.1
.\" (c) 2014 Richard Kettlewell
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.TH run-repeatedly 1
.SH NAME
run-repeatedly \- run and restart programs
.SH SYNOPSIS
.B run-repeatedly
.RI [ options ]
.RB [ -- ]
.I directory
.SH DESCRIPTION
\fBrun-repeatedly\fR runs one or more programs and restarts them when
they terminate.
.PP
The programs to be run are placed in \fIdirectory\fR.  Only executable
files with names that consist only of letters, digits, underscores and hyphens are
executed; everything else (including dot-files and backup files) is
ignored.
.PP
The directory is rescanned periodically.  If a new file appears in it
then it will be executed.  If a file disappears from it then the
corresponding program won't be stopped immediately; however it won't
be restarted next time it terminates, unless the \fB--shutdown\fR
option is used.
.PP
If any of SIGINT, SIGTERM or SIGHUP are received then all children are
signalled (with SIGTERM if SIGINT or SIGTERM was received, or with
SIGHUP if that was received).  Only when all children exit does
\fBrun-repeatedly\fR itself terminate.
.SH OPTIONS
.TP
\fB-r\fR \fIseconds\fR, \fB--rescan-interval\fR \fIseconds\fR
The minimum number of seconds to leave between each rescan of the
directory.  The default is 5 minutes.
.TP
\fB-e\fR \fIseconds\fR, \fB--execute-interval\fR \fIseconds\fR
The minimum number of seconds to leave between each restart of each
program. The default is 1 minute.
.TP
\fB-s\fR, \fB--shutdown\fR
When a program disappears from the control directory, attempt to shut
it down by sending it a SIGTERM.
.TP
\fB-d\fR, \fB--debug\fR
Debug mode.
\fB-h\fR, \fB--help\fR
Show summary of options.
.TP
\fB-V\fR, \fB--version\fR
Show version of program.
.SH AUTHOR
Richard Kettlewell <rjk@greenend.org.uk>


ChangeLog
2005-02-07 17:17:21 +0000  Richard Kettlewell  <rjk@greenend.org.uk>

        * adverbio.c: -O FD:PATH where the open(PATH, ...) returned a value
          matching FD would unhelpfuly close FD.  Problem reported by Colin
          Watson.
2005-02-07 17:14:08 +0000  Richard Kettlewell  <rjk@greenend.org.uk>

        * with-lock.c: bogus string literal
2003-07-19 17:35:12  Richard Kettlewell  <rjk@greenend.org.uk>

        * check-tested-everything, test-*: check that every executable has
          been tested
        
        * configure.ac: quieten autoconf
        
        * debug.c, utils.h, run-repeatedly.c: new debugging interface
        
        * test-run-repeatedly: fix test
        
2003-07-19 11:26:29  Richard Kettlewell  <rjk@greenend.org.uk>

        * prepare: don't configure if asked not to
        
        * test-run-repeatedly: tests
        
2003-03-08  Richard Kettlewell  <rjk@greenend.org.uk>

	* inplace.1: added some examples
	
2002-11-28  Richard Kettlewell  <rjk@greenend.org.uk>

	* iobuffer.c: use a single big circular buffer rather than lots of
	small buffers.  This makes more predictable and efficient use of
	memory than the previous strategy.

	Also change --max-buffer/-m to --buffer/-b since it is now the
	exact buffer size.

2002-11-23  Richard Kettlewell  <rjk@greenend.org.uk>

	* More test infrastructure.  Various tests improved.

	* run-repeatedly.c: underscore in filenames is allowed (to match
	run-parts)

2002-11-22  Richard Kettlewell  <rjk@greenend.org.uk>

	* various changes to build on FreeBSD.  daemon (at least) doesn't
	seem to work properly there though.

2002-11-21  Richard Kettlewell  <rjk@greenend.org.uk>

	* alarm.c: report failed setpgid correctly

	* lots of <syscall>_e functions to keep error handling out of the
	main code.  Convert a couple of malloc/realloc calls to the x-
	versions too (for the same reason).

	* more tests

2002-11-20  Richard Kettlewell  <rjk@greenend.org.uk>

	* test-daemon: more tests, including for some of the new features

	* logfds.c: adjust option letters so daemon & logfds are
	consistent.

	* daemon.c: saner error output handling - essentially build part
	of logfds into daemon.

2002-11-18  Richard Kettlewell  <rjk@greenend.org.uk>

	* anagrams.c: use dynamic memory allocation rather than fixed-size
	buffers.  Now there should be no size limits other than available
	memory.

	Also, now you can specify multiple word lists, and duplicate words
	don't cause duplicate anagrams.

2002-11-17  Richard Kettlewell  <rjk@greenend.org.uk>

	* new programs: iobuffer, anagrams

2002-11-11  Richard Kettlewell  <rjk@greenend.org.uk>

	* inplace.c: fix message reported when a subprocess gets a signal

	* alarm: new --verbose option

	* new programs: logfds, bind/accept/connect-socket, pidfile,
	run-as, daemon, run-repeatedly, with-lock

	* switch to new autoconf &c

2001-09-09  Richard Kettlewell  <rjk@greenend.org.uk>

	* configure.in: changed version to 0.6.0

	* added "daemon" program.

2001-08-12  Richard Kettlewell  <rjk@greenend.org.uk>

	* merged "inplace" and "adverbio" into a single package



run-repeatedly.c
/* 
   run-repeatedly - run and restart programs

   Copyright (C) 2001, 2003 Richard Kettlewell

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

#include <config.h>

#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <errno.h>
#include <time.h>
#include <ctype.h>
#include <signal.h>
#include <unistd.h>
#include <fcntl.h>
#include <getopt.h>
#include <dirent.h>
#include <sys/wait.h>
#include <sys/stat.h>

#include "utils.h"

struct daemon {
  struct daemon *next;			/* next daemon */
  char *path;				/* path */
  pid_t pid;				/* process ID or -1 */
  int present;				/* true if still present */
  int shutting_down;			/* true if already sent a
					 * shutdown signal */
  time_t last;				/* last start time */
};

static struct daemon *daemons;		/* list of managed daemons */
static int rescan_interval = 5*60;	/* interval between rescans */
static int execute_interval = 1*60;	/* interval between executions */
static const char *directory;		/* directory of daemons */
static int shutting_down;		/* shutting down? */
static int shutdown_obsolete;		/* terminate obsolete children */

/* option flags and variables */

static struct option const long_options[] =
{
  { "rescan-interval", required_argument, 0, 'i' },
  { "execute-interval", required_argument, 0, 'e' },
  { "shutdown", no_argument, 0, 's' },
  { "debug", no_argument, 0, 'd' },
  { "help", no_argument, 0, 'h' },
  { "version", no_argument, 0, 'V' },
  { 0, 0, 0, 0}
};

/* write a usage message to FP and exit with the specified status */

static void __attribute__((noreturn)) usage(FILE *fp, int exit_status) {
  if(fputs(
"Usage:\n"
"  run-repeatedly [options] [--] directory\n"
"\n"
"Options:\n"
"  -r SECONDS, --rescan-interval SECONDS   Set rescan interval\n"
"  -e SECONDS, --execute-interval SECONDS  Set execution interval\n"
"  -s, --shutdown                          Shut down children on removal\n"
"  -h, --help                              Usage message\n"
"  -V, --version                           Version number\n"
, fp) < 0)
    fatale("output error");
  exit(exit_status);
}

/* limit events to once per INTERVAL seconds.  TIMEP should point to a
 * time_t which will be used to track when events occur; it should be
 * initialized to 0 before the first call.
 *
 * If the event is allowed, *TIMEP will be updated and 0 will be
 * returned.
 *
 * If the event is not allowed yet, 1 will be returned.
 */

static int ratelimit(int interval, time_t *timep) {
  time_t now;

  time(&now);
  if(*timep == 0
     || now >= *timep + interval) {
    *timep = now;
    return 0;
  } else {
    debug("rate limited");
    return 1;
  }
}

/* rescan the directory, return nonzero if there is a serious error */

static int rescan(void) {
  DIR *dp;
  struct dirent *de;
  struct daemon *d;
  
  static time_t last;

  debug("rescanning");
  if(ratelimit(rescan_interval, &last))
    return 0;
  
  for(d = daemons; d; d = d->next)
    d->present = 0;
  if(!(dp = opendir(directory))) {
    errore("error opening directory %s", directory);
    return -1;
  }
  while((de = readdir(dp))) {
    int i;
    char *fullpath;
    struct stat sb;
    struct daemon *d;

    for(i = 0; de->d_name[i]; ++i)
      if(!(isalnum((unsigned char)de->d_name[i])
	   || de->d_name[i] == '_'
	   || de->d_name[i] == '-'))
	break;
    if(de->d_name[i])
      continue;
    fullpath = xstrdupcat3(directory, "/", de->d_name);
    if(stat(fullpath, &sb) < 0) {
      errore("cannot stat %s", fullpath);
      free(fullpath);
      continue;
    }
    if(!S_ISREG(sb.st_mode)
       || ((sb.st_mode & 0111) == 0))
      continue;
    for(d = daemons; d; d = d->next)
      if(!strcmp(d->path, fullpath))
	break;
    if(!d) {
      d = xmalloc(sizeof *d);
      d->next = daemons;
      d->path = fullpath;
      d->pid = -1;
      d->present = 1;
      d->last = 0;
      d->shutting_down = 0;
      daemons = d;
    } else {
      d->present = 1;
      free(fullpath);
    }
  }
  closedir(dp);
  if(debugging)
    for(d = daemons; d; d = d->next)
      if(d->present)
	debug("found %s", d->path);
  return 0;
}

/* restart daemon D */

static void restart(struct daemon *d) {
  pid_t pid;
  sigset_t ss;

  /* don't start the daemon more than once per minute */
  debug("restarting %s", d->path);
  if(ratelimit(execute_interval, &d->last))
    return;
  switch(pid = fork()) {
  case -1:
    errore("[%s] error calling fork", d->path);
    break;
  case 0:
    alarm(0);
    exiter = _exit;
    /* re-enable signals */
    sigemptyset(&ss);
    sigprocmask_e(SIG_SETMASK, &ss, 0);
    if(setpgid(0, 0) < 0)
      fatale("[%s] error calling setpgid", d->path);
    execl(d->path, d->path, (const char *)0);
    fatale("error executing %s", d->path);
  default:
    d->pid = pid;
    break;
  }
}

/* start any daemons that aren't running */

static void startall(void) {
  struct daemon *d;

  if(!shutting_down) {
    /* restart any dead daemons */
    for(d = daemons; d; d = d->next)
      if(d->present && d->pid == (pid_t)-1)
	restart(d);
  }
}

/* clean up any dead, unwanted daemons */

static void cleanup(void) {
  struct daemon **dd, *d;

  dd = &daemons;
  while((d = *dd)) {
    if(d->pid == (pid_t)-1 && (!d->present || shutting_down)) {
      debug("cleaning up %s", d->path);
      *dd = d->next;
      free(d->path);
      free(d);
    } else
      dd = &d->next;
  }
}

static void alarm_handler(int __attribute__((unused)) sig) {
  debug("SIGALRM");
      
  /* spot changes to the control directory */
  rescan();
  if(shutdown_obsolete) {
    /* zap any obsolete daemons */
    struct daemon *d;

    for(d = daemons; d; d = d->next)
      if(d->pid != (pid_t)-1
	 && !d->present
	 && !d->shutting_down) {
	debug("shutting down %s", d->path);
	if(kill(d->pid, SIGTERM) < 0)
	  errore("error sending signal %d (%s) to %s[%lu]",
		 SIGTERM, strsignal(SIGTERM),
		 d->path, (unsigned long)d->pid);
	d->shutting_down = 1;
      }
  }
  /* start any children that have terminated */
  startall();
  /* cleanup memory allocation for obsolete daemons */
  cleanup();
  /* set another alarm */
  alarm(1);
}

static void child_handler(int __attribute__((unused)) sig) {
  pid_t p;
  int w;

  debug("SIGCHLD");

  /* pick up any terminated children */
  while((p = waitpid(-1, &w, WNOHANG)) > 0) {
    struct daemon *d;
    
    /* find which child terminated, if any */
    for(d = daemons; d && d->pid != p; d = d->next)
      ;
    if(d) {
      error("%s[%lu]: %s", d->path, (unsigned long)p, wstat(w));
      d->pid = -1;
    } else
      error("unknown process %lu: %s", (unsigned long)p, wstat(w));
  }
  if(p < 0) {
    switch(errno) {
    case EINTR:
    case ECHILD:
      debug("ignored error calling waitpid");
      break;
    default:
      errore("error calling waitpid");
      break;
    }
  } else if(p == 0)
    debug("waitpid returned 0");
  /* restart any terminated children */
  cleanup();
  startall();
  /* terminate if we reaped the last child */
  if(daemons == 0 && shutting_down) {
    sigset_t ss;

    debug("terminating myself");
    if(signal(shutting_down, SIG_DFL) == SIG_ERR)
      fatale("error calling signal");
    if(kill(getpid(), shutting_down) < 0)
      fatale("error sending signal %d (%s) to self [%lu]",
	     shutting_down, strsignal(shutting_down),
	     (unsigned long)getpid());
    sigemptyset(&ss);
    sigaddset(&ss, shutting_down);
    sigprocmask_e(SIG_UNBLOCK, &ss, 0);
    fatal("unexpectedly failed to terminate on signal %d to self",
	  shutting_down);
  }
}

static void shutdown_handler(int sig) {
  struct daemon *d;
  int s;
  
  /* record which signal we actually caught */
  shutting_down = sig;
  debug("%s", strsignal(shutting_down));
  /* which signal to actually send to children.  We substitute
   * SIGTERM for SIGINT as some programs e.g. bash don't seem to
   * like to terminate on SIGINT. */
  s = shutting_down == SIGINT ? SIGTERM : shutting_down;
  /* send the signal */
  for(d = daemons; d; d = d->next)
    if(d->pid != (pid_t)-1
       && !d->shutting_down) {
      debug("sending signal %d to %lu",
	    s, (unsigned long)d->pid);
      if(kill(d->pid, s) < 0)
	errore("error sending signal %d (%s) to %s[%lu]",
	       s, strsignal(s),
	       d->path, (unsigned long)d->pid);
    }
  /* turn off periodic rescans */
  alarm(0);
  /* do a speculative waitpid */
  child_handler(0);
}

int main(int argc, char **argv) {
  int n;

  setprogname(argv[0]);

  while((n = getopt_long(argc, argv, 
			 "hVr:e:ds",
			 long_options, (int *)0)) >= 0) {
    switch(n) {
    case 'V':
      printf("run-repeatedly %s\n", VERSION);
      return 0;

    case 'h':
      usage(stdout, 0);
      return 0;

    case 'r':
      rescan_interval = atoi(optarg);
      break;

    case 'e':
      execute_interval = atoi(optarg);
      break;

    case 'd':
      ++debugging;
      break;

    case 's':
      shutdown_obsolete = 1;
      break;
      
    default:
      usage(stderr, 1);
    }
  }

  /* pick up the directory */
  if(optind >= argc)
    fatal("no directory specified");
  directory = argv[optind++];

  if(optind < argc)
    fatal("unexpected extra arguments");

  /* set up signal handling */
  handle_signal(SIGCHLD, child_handler);
  handle_signal(SIGALRM, alarm_handler);
  handle_signal(SIGTERM, shutdown_handler);
  handle_signal(SIGINT, shutdown_handler);
  handle_signal(SIGHUP, shutdown_handler);
  
  /* perform initial scan */
  if(rescan())
    exit(-1);

  /* start up daemons */
  startall();

  /* set up alarms */
  alarm(1);

  /* enter event loop */
  signal_loop();
  
}

/*
Local Variables:
c-basic-offset:2
comment-column:40
End:
*/
